File: app/src/main/java/com/ddorocare/brand_audit/LoginActivity.java
package com.ddorocare.brand_audit;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

public class LoginActivity extends AppCompatActivity {

    private View _bg__login_page;
    private ImageView logo_;
    private View _bg__nama_pengguna_atau_email_ek1;
    private View rectangle_6;
    private TextView email;
    private View _bg__kata_sandi_ek1;
    private View rectangle_7;
    private TextView kata_sandi_ek2;
    private TextView lupa_kata_sandi_;
    private View _bg__sign_in_ek1;
    private View rectangle_8;
    private TextView sign_in_ek2;
    private TextView login;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        _bg__login_page = (View) findViewById(R.id._bg__login_page);
        logo_ = (ImageView) findViewById(R.id.logo_);
        _bg__nama_pengguna_atau_email_ek1 = (View) findViewById(R.id._bg__nama_pengguna_atau_email_ek1);
        rectangle_6 = (View) findViewById(R.id.rectangle_6);
        email = (TextView) findViewById(R.id.email);
        _bg__kata_sandi_ek1 = (View) findViewById(R.id._bg__kata_sandi_ek1);
        rectangle_7 = (View) findViewById(R.id.rectangle_7);
        kata_sandi_ek2 = (TextView) findViewById(R.id.kata_sandi_ek2);
        lupa_kata_sandi_ = (TextView) findViewById(R.id.lupa_kata_sandi_);
        _bg__sign_in_ek1 = (View) findViewById(R.id._bg__sign_in_ek1);
        rectangle_8 = (View) findViewById(R.id.rectangle_8);
        sign_in_ek2 = (TextView) findViewById(R.id.sign_in_ek2);
        login = (TextView) findViewById(R.id.login);
    }
}

File: app/src/main/java/com/ddorocare/brand_audit/LoginActivity.kt
package com.ddorocare.brand_audit

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStore
import androidx.lifecycle.ViewModelProvider
import com.ddorocare.brand_audit.model.LoginRequest
import com.ddorocare.brand_audit.model.UserModel
import com.ddorocare.brand_audit.model.UserPreference
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText


private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "settings")

class LoginActivity : AppCompatActivity() {
    //    private lateinit var binding: ActivityLoginBinding
    private lateinit var loginViewModel: LoginViewModel
    private lateinit var email: TextInputEditText
    private lateinit var password: TextInputEditText
    private lateinit var loginButton: MaterialButton
    private lateinit var registerButton: MaterialButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(R.layout.activity_login)
        email = findViewById(R.id.edt_email)
        password = findViewById(R.id.edt_password)
        loginButton = findViewById(R.id.btn_login)
        registerButton = findViewById(R.id.btn_register)
        setupViewModel()
        setupAction()




        registerButton.setOnClickListener {
            startActivity(Intent(this, RegisterActivity::class.java))
        }

    }


    private fun setupViewModel() {
        loginViewModel = ViewModelProvider(
            this,
            ViewModelFactory(UserPreference.getInstance(dataStore))
        )[LoginViewModel::class.java]

        loginViewModel.logout()
    }

    private fun setupAction() {
        loginButton.setOnClickListener {
            val email = email.text.toString()
            val password = password.text.toString()

            loginViewModel.login(LoginRequest(email, password)).observe(this) { result ->
                if (result != null) {
                    when (result) {
                        is ResultCustom.Loading -> findViewById<ProgressBar>(R.id.pb_loading).visibility =
                            View.VISIBLE

                        is ResultCustom.Success -> {
                            val user = result.data
                            findViewById<ProgressBar>(R.id.pb_loading).visibility = View.GONE
                            loginViewModel.saveLogin(user.token)
                            loginViewModel.saveUser(
                                UserModel(
                                    result.data.token,
                                    result.data.fullName!!,
                                    result.data.username!!,
                                    true,
                                    role = result.data.role!!
                                )
                            )
                            AlertDialog.Builder(this).apply {
                                val intent = Intent(context, MainActivity::class.java)
                                intent.putExtra("name", result.data.fullName)
                                intent.putExtra("role", result.data.role)
                                Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
                                startActivity(intent)
                                finish()
                            }

                        }

                        is ResultCustom.Error -> {
                            findViewById<ProgressBar>(R.id.pb_loading).visibility = View.GONE
                            Toast.makeText(
                                this@LoginActivity,
                                result.error,
                                Toast.LENGTH_SHORT
                            ).show()
                        }
                    }
                }
            }
        }
    }

}
